odbc/src/api/data.rs

```diff
@@ -196,9 +198,58 @@ impl IntoCData for Bson {
     fn to_datetime(&self) -> Result<DateTime<Utc>> {
         match self {
             Bson::DateTime(d) => Ok((*d).into()),
-            Bson::String(s) => Utc
-                .datetime_from_str(s, "%FT%H:%M:%S%.3fZ")
-                .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone())),
+            Bson::String(s) => {
+                if s.contains('T') {
+                    // received a datetime string
+                    Utc.datetime_from_str(s, "%FT%H:%M:%S%.3fZ")
+                        .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))
+                } else if s.contains('-') {
+                    // received a date string
+                    let date = NaiveDate::parse_from_str(s, "%F")
+                        .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))?;
+                    Ok(DateTime::<Utc>::from_utc(date.and_hms(0, 0, 0), Utc))
+                } else {
+                    // received a time string -- should return utc today + time
+                    let time = NaiveTime::parse_from_str(s, "%H:%M:%S%.3f")
+                        .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))?;
+                    Ok(DateTime::<Utc>::from_utc(
+                        NaiveDateTime::new(Utc::today().naive_utc(), time),
+                        Utc,
+                    ))
+                }
+            }
+            o => Err(ODBCError::RestrictedDataType(o.to_type_str(), DATETIME)),
+        }
+    }
+
+    fn to_date(&self) -> Result<NaiveDate> {
+        match self {
+            Bson::DateTime(d) => Ok((*d).to_chrono().date_naive()),
+            Bson::String(s) => {
+                let fmt = if s.contains('T') {
+                    "%FT%H:%M:%S%.3fZ"
+                } else {
+                    "%F"
+                };
+                NaiveDate::parse_from_str(s, fmt)
+                    .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))
+            }
+            o => Err(ODBCError::RestrictedDataType(o.to_type_str(), DATETIME)),
+        }
+    }
+
+    fn to_time(&self) -> Result<NaiveTime> {
+        match self {
+            Bson::DateTime(d) => Ok((*d).to_chrono().time()),
+            Bson::String(s) => {
```

Comment:
Similar to the date case above, it seems that the truncation warning is not handled.
We should return SUCCESS_WITH_INFO-Truncated data when the `data value is a valid timestamp-value; fractional seconds portion is nonzero`. Note that if I read the spec correctly, when the value is a time and the fractional seconds portion is nonzero, it apparently should fail instead of being a truncation error 😵‍💫

Reply:
yeah, this is my interpretation too; added in the coming updates